### ISSUE TITLE
How to Unmarshal back to int64 in the right way using custom Unmarshal implementation?

### ISSUE THREAD

**Issue by bxcodec** [vmihailenco/msgpack#195]
I'm making a custom Unmarshal implementation for my own projects. It's more look like this below. 
I have a struct that has `[]int64`, so in custom unmarshal, I'm planning to Unmarshal it first to map[string]interface{} then cast by its original data type. But when I tried to cast the  `[]int64` it throws panic. 

```go
package main

import (
	"log"

	"github.com/vmihailenco/msgpack"
)

type SampleStruct struct {
	ArrayInt64 []int64 `msgpack:"data"`
        OtherTypes interface{} // the reason why using custom Unmarshal with map[string]interface{}
        // other data types and struct/interface
}

func (s *SampleStruct) UnmarshalMsgpack(data []byte) error {

	temp := map[string]interface{}{}

	err := msgpack.Unmarshal(data, &temp)
	if err != nil {
		return err
	}

	arrInterface := temp["data"].([]interface{})
	arr := []int64{}
	for _, item := range arrInterface {
		arr = append(arr, item.(int64)) // This line throw panic
	}
	s.ArrayInt64 = arr
	return err
}

func main() {
	sample := &SampleStruct{
		ArrayInt64: []int64{1, 2, 4, 5},
	}

	msbyt, err := msgpack.Marshal(sample)
	if err != nil {
		log.Fatal(err)
	}

	res := &SampleStruct{}
	err = msgpack.Unmarshal(msbyt, res)
	log.Fatal(err)
}

```

The panic messages said that the data type is `int8` not `int64`
```shell
panic: interface conversion: interface {} is int8, not int64
```

**Comment by bxcodec** [comment]
Currently, I'm using a function to manage the convertion : 
```go
func convertToInt64(data interface{}) int64 {

	int64Type := reflect.TypeOf(int64(0))
	v := reflect.ValueOf(data)
	v = reflect.Indirect(v)
	if !v.Type().ConvertibleTo(int64Type) {
		return 0
	}
	res := v.Convert(int64Type)
	return res.Int()
}
```
Any better idea?

**Comment by vmihailenco** [comment]
When you decode into interface{} there can be anything - you can't assume specific type. In your case numbers are *encoded* as msgpack int8 to save space because they fit into int8. Probably https://godoc.org/github.com/vmihailenco/msgpack#Decoder.UseDecodeInterfaceLoose will be helpful.

**Comment by vmihailenco** [comment]
This should be fixed in v3.4 - encoder should save underlying type.
